File: java/client/src/main/java/glide/api/models/pool/ClientPool.java
/** Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0 */
package glide.api.models.pool;

import static connection_request.ConnectionRequestOuterClass.*;

import glide.api.BaseClient;
import glide.api.models.configuration.BackoffStrategy;
import glide.api.models.configuration.BaseClientConfiguration;
import glide.api.models.configuration.GlideClusterClientConfiguration;
import glide.api.models.configuration.ServerCredentials;
import glide.api.models.exceptions.ClosingException;
import glide.ffi.resolvers.GlidePoolResolver;
import glide.internal.GlideNativeBridge;
import java.time.Duration;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * Client-instance pool backed by a Rust-side pool managing CoreClient lifecycle.
 *
 * <p>Callers borrow a client via {@link #acquire()}, use it for commands, and return it
 * via {@code close()} on the borrowed client. The pool handles creation, reuse, health
 * checking, idle eviction, and graceful shutdown.
 *
 * <p>The Rust pool owns the actual state (LIFO idle list, bounded size, eviction task).
 * This Java class provides acquire-with-timeout semantics and the PooledClient proxy pattern.
 *
 * @param <T> the client type (GlideClient or GlideClusterClient)
 */
public class ClientPool<T extends BaseClient> implements AutoCloseable {

    /** Pool states */
    private static final int RUNNING = 0;
    private static final int CLOSING = 1;
    private static final int CLOSED = 2;

    /** Opaque handle to the Rust-side pool. */
    private final long poolId;

    /** Pool configuration (immutable after creation). */
    private final ClientPoolConfig config;

    /** Factory for wrapping client_id into Java proxy type. */
    private final ClientFactory<T> factory;

    /** Pool lifecycle state. */
    private final AtomicInteger state = new AtomicInteger(RUNNING);

    /**
     * Cache of Java client wrappers keyed by client_id.
     * The factory is called once per client_id (on first acquire of that id).
     * Subsequent borrows of the same client_id reuse the cached wrapper — no allocation.
     */
    private final java.util.concurrent.ConcurrentHashMap<Long, T> clientCache =
            new java.util.concurrent.ConcurrentHashMap<>();

    /**
     * Private constructor — use {@link #create} factory method.
     */
    private ClientPool(long poolId, ClientPoolConfig config, ClientFactory<T> factory) {
        this.poolId = poolId;
        this.config = config;
        this.factory = factory;
    }

    /**
     * Creates a new client pool backed by the Rust core.
     *
     * <p>This method:
     * <ol>
     *   <li>Validates the config</li>
     *   <li>Serializes the ConnectionRequest to protobuf bytes</li>
     *   <li>Calls glide_pool_create via JNI to create the Rust pool</li>
     *   <li>Returns a ClientPool instance in RUNNING state</li>
     * </ol>
     *
     * @param config pool configuration
     * @param factory client proxy factory
     * @param <T> the client type
     * @return a new ClientPool
     * @throws IllegalArgumentException if config is invalid
     * @throws RuntimeException if Rust pool creation fails
     */
    public static <T extends BaseClient> ClientPool<T> create(
            ClientPoolConfig config, ClientFactory<T> factory) {
        // Validate config
        config.validate();

        // Serialize ConnectionRequest to protobuf bytes
        // TODO: Implement proper serialization using ConnectionManager's protobuf building logic
        // For now, use a placeholder — the actual serialization will be wired in integration
        byte[] connectionRequestBytes = serializeConnectionRequest(config.getClientConfig());

        // Create the Rust pool
        long poolId = GlidePoolResolver.glidePoolCreate(
                config.getMaxSize(),
                config.getMinIdle(),
                config.getIdleTimeout().toMillis(),
                config.getRequestTimeout().toMillis(),
                connectionRequestBytes);

        if (poolId == -1) {
            throw new IllegalArgumentException("Invalid pool configuration rejected by Rust core");
        }
        if (poolId < 0) {
            throw new RuntimeException(
                    "Failed to create pool in Rust core (error code: " + poolId + ")");
        }

        return new ClientPool<>(poolId, config, factory);
    }

    /**
     * Get the Rust pool handle.
     * @return the pool_id
     */
    public long getPoolId() {
        return poolId;
    }

    /**
     * Get the pool configuration.
     * @return the config
     */
    public ClientPoolConfig getConfig() {
        return config;
    }

    /**
     * Get the client factory.
     * @return the factory
     */
    public ClientFactory<T> getFactory() {
        return factory;
    }

    /**
     * Check if the pool is running (accepting acquire requests).
     * @return true if running
     */
    public boolean isRunning() {
        return state.get() == RUNNING;
    }

    /**
     * Check if the pool has been closed.
     * @return true if closed
     */
    public boolean isClosed() {
        return state.get() == CLOSED;
    }

    // ========== Pool metrics ==========

    /**
     * Get the number of idle clients in the pool.
     * @return idle count
     */
    public int getIdleCount() {
        int[] metrics = GlidePoolResolver.glidePoolMetrics(poolId);
        return metrics != null ? metrics[0] : 0;
    }

    /**
     * Get the number of actively borrowed clients.
     * @return active count
     */
    public int getActiveCount() {
        int[] metrics = GlidePoolResolver.glidePoolMetrics(poolId);
        return metrics != null ? metrics[1] : 0;
    }

    /**
     * Get the total number of clients (idle + active + cleaning + creating).
     * @return total count
     */
    public int getTotalCount() {
        int[] metrics = GlidePoolResolver.glidePoolMetrics(poolId);
        return metrics != null ? metrics[2] : 0;
    }

    /**
     * Get the configured max pool size.
     * @return max size
     */
    public int getMaxSize() {
        return config.getMaxSize();
    }

    // ========== Acquire / Release ==========

    /**
     * Acquire a client from the pool with the default timeout.
     *
     * @return a CompletableFuture that resolves to a borrowed client proxy
     */
    public CompletableFuture<T> acquire() {
        return acquire(config.getAcquireTimeout());
    }

    /**
     * Acquire a client from the pool with a custom timeout.
     *
     * <p>Internally retries {@code glidePoolTryAcquire} with exponential backoff
     * (1ms initial, 50ms cap) until a client_id is obtained or the timeout expires.
     *
     * @param timeout maximum time to wait
     * @return a CompletableFuture that resolves to a borrowed client proxy
     */
    public CompletableFuture<T> acquire(Duration timeout) {
        if (state.get() != RUNNING) {
            CompletableFuture<T> f = new CompletableFuture<>();
            f.completeExceptionally(new ClosingException("Pool is shutting down"));
            return f;
        }

        return CompletableFuture.supplyAsync(() -> {
            long deadlineNanos = System.nanoTime() + timeout.toNanos();
            long backoffMs = 1;

            while (System.nanoTime() < deadlineNanos) {
                // Check state on each iteration
                if (state.get() != RUNNING) {
                    throw new RuntimeException(new ClosingException("Pool is shutting down"));
                }

                long clientId = GlidePoolResolver.glidePoolTryAcquire(poolId);

                if (clientId >= 0) {
                    // Success — return cached wrapper or create on first use
                    T cached = clientCache.get(clientId);
                    if (cached != null) {
                        return cached;
                    }
                    return clientCache.computeIfAbsent(clientId,
                            id -> factory.wrapClientId(id, poolId));
                }

                if (clientId == -2) {
                    // Invalid pool_id — pool was destroyed
                    throw new RuntimeException(new ClosingException("Pool has been destroyed"));
                }

                // clientId == -1: pool exhausted or lock contended, backoff and retry
                long remainingMs = (deadlineNanos - System.nanoTime()) / 1_000_000;
                long sleepMs = Math.min(backoffMs, Math.max(remainingMs, 0));
                if (sleepMs <= 0) {
                    break;
                }
                try {
                    Thread.sleep(sleepMs);
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    throw new RuntimeException("Acquire interrupted", e);
                }
                backoffMs = Math.min(backoffMs * 2, 50);
            }

            throw new RuntimeException(new java.util.concurrent.TimeoutException(
                    "Pool exhausted: could not acquire client within " + timeout));
        });
    }

    /**
     * Release a borrowed client back to the pool. Called by PooledClient.close().
     *
     * @param clientId the Rust-assigned client_id to release
     */
    public void release(long clientId) {
        GlidePoolResolver.glidePoolRelease(poolId, clientId);
    }

    /**
     * Get or create a cached Java client wrapper for the given client_id.
     * The factory is called once per client_id; subsequent calls return the cached instance.
     * This is the fast path used by acquire() to avoid allocation on every borrow.
     *
     * @param clientId the Rust-assigned client_id
     * @return the cached client wrapper
     */
    public T getOrCreateClient(long clientId) {
        // Fast path: plain get() avoids bucket locking on cache hit
        T cached = clientCache.get(clientId);
        if (cached != null) {
            return cached;
        }
        // Slow path (first borrow of this client_id): allocate wrapper
        return clientCache.computeIfAbsent(clientId, id -> factory.wrapClientId(id, poolId));
    }

    // ========== Lifecycle ==========

    @Override
    public void close() {
        close(Duration.ofSeconds(30));
    }

    /**
     * Gracefully shut down the pool with a custom grace period.
     *
     * @param gracePeriod maximum time to wait for borrowed clients to be returned
     */
    public void close(Duration gracePeriod) {
        if (!state.compareAndSet(RUNNING, CLOSING)) {
            return; // Already closing or closed
        }
        // TODO: Wait for outstanding borrowed clients (task 6.3)
        GlidePoolResolver.glidePoolDestroy(poolId);
        state.set(CLOSED);
    }

    /**
     * Immediately destroy the pool without waiting for borrowed clients.
     */
    public void closeImmediately() {
        state.set(CLOSED);
        GlidePoolResolver.glidePoolDestroy(poolId);
    }

    // ========== Internal helpers ==========

    /**
     * Serialize a BaseClientConfiguration to protobuf ConnectionRequest bytes.
     *
     * <p>Builds a ConnectionRequest protobuf matching the logic in ConnectionManager.connectToValkey(),
     * covering the essential fields needed for the Rust side to create a real GlideClient.
     */
    private static byte[] serializeConnectionRequest(BaseClientConfiguration config) {
        ConnectionRequest.Builder requestBuilder = ConnectionRequest.newBuilder();

        // Add addresses
        for (glide.api.models.configuration.NodeAddress addr : config.getAddresses()) {
            NodeAddress nodeAddress =
                    NodeAddress.newBuilder()
                            .setHost(addr.getHost())
                            .setPort(addr.getPort())
                            .build();
            requestBuilder.addAddresses(nodeAddress);
        }

        // Set TLS mode
        if (config.isUseTLS()) {
            requestBuilder.setTlsMode(TlsMode.SecureTls);
        } else {
            requestBuilder.setTlsMode(TlsMode.NoTls);
        }

        // Set cluster mode
        boolean isCluster = config instanceof GlideClusterClientConfiguration;
        requestBuilder.setClusterModeEnabled(isCluster);

        // Set request timeout
        int requestTimeoutMs = config.getRequestTimeout() != null
                ? config.getRequestTimeout()
                : (int) GlideNativeBridge.getGlideCoreDefaultRequestTimeoutMs();
        requestBuilder.setRequestTimeout(requestTimeoutMs);

        // Set connection timeout (same as request timeout if not explicitly configured)
        requestBuilder.setConnectionTimeout(requestTimeoutMs);

        // Set inflight requests limit
        int inflightLimit = config.getInflightRequestsLimit() != null
                ? config.getInflightRequestsLimit()
                : GlideNativeBridge.getGlideCoreDefaultMaxInflightRequests();
        requestBuilder.setInflightRequestsLimit(inflightLimit);

        // Set authentication if credentials are present
        ServerCredentials credentials = config.getCredentials();
        if (credentials != null) {
            AuthenticationInfo.Builder authBuilder = AuthenticationInfo.newBuilder();
            if (credentials.getUsername() != null) {
                authBuilder.setUsername(credentials.getUsername());
            }
            if (credentials.getPassword() != null) {
                authBuilder.setPassword(credentials.getPassword());
            }
            requestBuilder.setAuthenticationInfo(authBuilder.build());
        }

        // Set read from strategy
        String readFromName = config.getReadFrom().name();
        if ("PRIMARY".equals(readFromName)) {
            requestBuilder.setReadFrom(ReadFrom.Primary);
        } else if ("PREFER_REPLICA".equals(readFromName)) {
            requestBuilder.setReadFrom(ReadFrom.PreferReplica);
        }

        // Set client name if provided
        if (config.getClientName() != null) {
            requestBuilder.setClientName(config.getClientName());
        }

        // Set database ID if provided
        if (config.getDatabaseId() != null) {
            requestBuilder.setDatabaseId(config.getDatabaseId());
        }

        // Set protocol version if specified
        if (config.getProtocol() != null) {
            if ("RESP2".equals(config.getProtocol().name())) {
                requestBuilder.setProtocol(ProtocolVersion.RESP2);
            } else if ("RESP3".equals(config.getProtocol().name())) {
                requestBuilder.setProtocol(ProtocolVersion.RESP3);
            }
        }

        // Set reconnect strategy if configured
        BackoffStrategy reconnectStrategy = config.getReconnectStrategy();
        if (reconnectStrategy != null) {
            ConnectionRetryStrategy.Builder retryBuilder = ConnectionRetryStrategy.newBuilder();
            if (reconnectStrategy.getNumOfRetries() != null) {
                retryBuilder.setNumberOfRetries(reconnectStrategy.getNumOfRetries());
            }
            if (reconnectStrategy.getFactor() != null) {
                retryBuilder.setFactor(reconnectStrategy.getFactor());
            }
            if (reconnectStrategy.getExponentBase() != null) {
                retryBuilder.setExponentBase(reconnectStrategy.getExponentBase());
            }
            requestBuilder.setConnectionRetryStrategy(retryBuilder.build());
        }

        return requestBuilder.build().toByteArray();
    }
}

File: python/glide-sync/glide_sync/client_pool.py
# Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0

"""
Client-instance pool for the synchronous GLIDE client (Feature 1).

Wraps the Rust-side pool exposed via FFI. Callers borrow a client from
the pool, use it for commands, and return it. The pool handles creation,
LIFO reuse, and bounded size.

Usage:
    from glide_sync import GlideClient, GlideClientConfiguration, NodeAddress
    from glide_sync.client_pool import ClientPool, PoolConfig

    config = GlideClientConfiguration([NodeAddress("localhost", 6379)])
    pool_config = PoolConfig(max_size=10, min_idle=2)
    pool = ClientPool(config, pool_config)

    client = pool.acquire()
    client.set("key", "value")
    result = client.get("key")
    pool.release(client)

    # Or use as context manager:
    with pool.borrow() as client:
        client.set("key", "value")
"""

import time
from dataclasses import dataclass, field
from typing import Optional

from glide_shared._glide_ffi import _GlideFFI
from glide_shared.config import BaseClientConfiguration, GlideClusterClientConfiguration

from .glide_client import BaseClient, GlideClient, GlideClusterClient


@dataclass
class PoolConfig:
    """Configuration for the client-instance pool."""

    max_size: int = 10
    """Maximum number of clients in the pool."""

    min_idle: int = 1
    """Minimum idle clients to pre-warm at creation."""

    idle_timeout_ms: int = 300_000
    """Evict idle clients after this duration (ms). Default: 5 minutes."""

    request_timeout_ms: int = 5_000
    """Request timeout for cleanup operations (ms)."""

    acquire_timeout_s: float = 5.0
    """Maximum time to wait when pool is exhausted (seconds)."""


class ClientPool:
    """
    Client-instance pool backed by the Rust core.

    The Rust pool owns client lifecycle (creation, health checks, eviction).
    This Python class provides the acquire-with-timeout retry loop and
    wraps raw client_id handles into usable GlideClient instances.
    """

    __slots__ = (
        '_ffi_instance', '_ffi', '_lib', '_client_config', '_pool_config',
        '_closed', '_client_cache', '_conn_req_bytes', '_pool_id',
    )

    def __init__(
        self,
        client_config: BaseClientConfiguration,
        pool_config: Optional[PoolConfig] = None,
    ):
        self._ffi_instance = _GlideFFI()
        self._ffi = self._ffi_instance.ffi
        self._lib = self._ffi_instance.lib
        self._client_config = client_config
        self._pool_config = pool_config or PoolConfig()
        self._closed = False

        # Clients cache: client_id → (GlideClient, cached_adapter_ptr)
        # The ffi.cast is cached so we don't redo it on every borrow.
        self._client_cache: dict = {}

        # Serialize the connection request protobuf
        is_cluster = isinstance(client_config, GlideClusterClientConfiguration)
        conn_req = client_config._create_a_protobuf_conn_request(cluster_mode=is_cluster)
        self._conn_req_bytes = conn_req.SerializeToString()

        # Create the Rust pool
        pool_id = self._lib.glide_pool_create(
            self._pool_config.max_size,
            self._pool_config.min_idle,
            self._pool_config.idle_timeout_ms,
            self._pool_config.request_timeout_ms,
            self._conn_req_bytes,
            len(self._conn_req_bytes),
        )

        if pool_id == -1:
            raise ValueError("Invalid pool configuration")
        if pool_id < 0:
            raise RuntimeError(f"Failed to create pool (error code: {pool_id})")

        self._pool_id = pool_id

    @property
    def pool_id(self) -> int:
        return self._pool_id

    @property
    def is_closed(self) -> bool:
        return self._closed

    def acquire(self, timeout: Optional[float] = None) -> int:
        """
        Acquire a client_id from the pool.

        Retries with exponential backoff until a client is available or timeout expires.

        Args:
            timeout: Max seconds to wait. Defaults to pool_config.acquire_timeout_s.

        Returns:
            client_id (positive integer) for command dispatch.

        Raises:
            TimeoutError: If no client available within timeout.
            RuntimeError: If pool is closed or invalid.
        """
        if self._closed:
            raise RuntimeError("Pool is closed")

        timeout = timeout if timeout is not None else self._pool_config.acquire_timeout_s
        deadline = time.monotonic() + timeout
        backoff_ms = 1.0

        while time.monotonic() < deadline:
            client_id = self._lib.glide_pool_try_acquire(self._pool_id)

            if client_id >= 0:
                return client_id

            if client_id == -2:
                raise RuntimeError("Invalid pool_id — pool was destroyed")

            # Pool exhausted, backoff and retry
            remaining = deadline - time.monotonic()
            sleep_s = min(backoff_ms / 1000.0, max(remaining, 0))
            if sleep_s <= 0:
                break
            time.sleep(sleep_s)
            backoff_ms = min(backoff_ms * 2, 50.0)

        raise TimeoutError(
            f"Pool exhausted: could not acquire client within {timeout}s"
        )

    def release(self, client_id: int) -> None:
        """Release a borrowed client back to the pool."""
        self._lib.glide_pool_release(self._pool_id, client_id)

    def borrow(self, timeout: Optional[float] = None):
        """
        Returns a context manager that acquires a client from the Rust pool.

        Usage:
            with pool.borrow() as client:
                client.set("key", "value")
                result = client.get("key")
        """
        return _BorrowContext(self, timeout)

    def _get_or_create_client(self, client_id: int) -> BaseClient:
        """
        Get a usable client wrapper for the given client_id.
        
        The Rust pool manages the actual GlideClient + ClientAdapter lifecycle.
        When a client_id is acquired, glide_pool_get_client_ptr() returns the
        ClientAdapter pointer that the existing command dispatch path uses.
        We create a thin Python wrapper that points to this native adapter.
        
        The wrapper and ffi.cast are cached per client_id — no allocation on
        subsequent borrows of the same client.
        """
        cached = self._client_cache.get(client_id)
        if cached is not None:
            return cached

        # Get the native ClientAdapter pointer from the pool
        adapter_ptr = self._lib.glide_pool_get_client_ptr(client_id)
        if adapter_ptr == 0:
            raise RuntimeError(
                f"Pool client_id {client_id} has no associated ClientAdapter pointer"
            )

        # Create a GlideClient shell that uses this adapter pointer for commands.
        # Cache the ffi.cast result so subsequent borrows pay zero FFI overhead.
        cast_ptr = self._ffi.cast("void*", adapter_ptr)

        client = GlideClient.__new__(GlideClient)
        client._ffi = self._ffi
        client._lib = self._lib
        client._config = self._client_config
        client._is_closed = False
        client._pubsub_queue = []
        client._pubsub_lock = __import__('threading').Lock()
        client._pubsub_condition = __import__('threading').Condition(client._pubsub_lock)
        client._pubsub_callback_ref = None
        # Set the core client pointer to the pooled adapter (cached cast)
        client._core_client = cast_ptr

        self._client_cache[client_id] = client
        return client

    def metrics(self) -> dict:
        """Get pool metrics: idle, active, total counts."""
        idle = self._ffi.new("uint32_t*")
        active = self._ffi.new("uint32_t*")
        total = self._ffi.new("uint32_t*")
        result = self._lib.glide_pool_metrics(self._pool_id, idle, active, total)
        if result != 0:
            return {"idle": 0, "active": 0, "total": 0}
        return {
            "idle": idle[0],
            "active": active[0],
            "total": total[0],
        }

    def close(self) -> None:
        """Destroy the pool and all its clients."""
        if not self._closed:
            self._closed = True
            self._lib.glide_pool_destroy(self._pool_id)
            # Don't close cached clients — glide_pool_destroy handles native cleanup
            self._client_cache.clear()

    def __enter__(self):
        return self

    def __exit__(self, *args):
        self.close()

    def __del__(self):
        if not self._closed:
            self.close()


class _BorrowContext:
    """Fast context manager for pool borrow/release (avoids generator overhead)."""

    __slots__ = ('_pool', '_timeout', '_client_id', '_client')

    def __init__(self, pool: 'ClientPool', timeout):
        self._pool = pool
        self._timeout = timeout
        self._client_id = -1
        self._client = None

    def __enter__(self):
        self._client_id = self._pool.acquire(self._timeout)
        self._client = self._pool._get_or_create_client(self._client_id)
        return self._client

    def __exit__(self, *_):
        self._pool.release(self._client_id)
        return False
